### ISSUE TITLE
Issues with "older" connections returning SCRAM-SHA-1 authentication failed

### ISSUE THREAD

**Issue by joningib** [Automattic/mongoose#4257]
We have a replicaset connection with our Node.js backend with mongoose 4.5.1 and mongodb 2.1.18 communicating with 3 data processes in a replica set running mongodb core server 3.2.7

For the past months we have updated mongoose a few times and mongodb as well. We started seeing increase in our db logs of the following errors:
`2016-06-20T14:39:14.212+0000 I ACCESS   [conn7980] SCRAM-SHA-1 authentication failed for XXX on YYY from client x.x.x.x ; BadValue: Unmatched SCRAM-SHA-1 nonce received from client in second step, expected ZZZ but received NNNN
`
These connections then fail authorisation on all subsequent queries using mongoose from our API 
`
`2016-06-21T07:00:00.279+0000 I ACCESS   [conn7980] Unauthorized: not authorized on XXXX to execute command`

 until, randomly, the connection is disconnected - which is due to some mongoose specifications regarding reconnect.... or something else (which I haven't managed to figure out yet.

However, this is having major affect on our production envornment and we cannot pinpoint where this is coming from. I THINK this might have something to due with a similar authentication error in mongodb native node.js driver as mentioned here:
https://jira.mongodb.org/browse/NODE-722

However, we have downgraded from 2.1.21 to 2.1.18 again due to the bugs mentioned, but still, this problem persists.

Our mongoose.CreateConnection connects to a replicaset with three processes defined in a connection string with the following options

readPref: Primary Preferred
w: 1
wtimeout: 10000

keepAlive: 1
connectionTimeoutMS: 45000
conectWithNoPrimary: true
poolSize: 50
ssl: true
sslCert
sslKey
sslpass

and then
username and pass for our authentication which is SCRAM-SHA-1

We are using cloud.mongodb to deploy and manage our mongodb solution, and used it to create all auth mechanism in our db, and setup replica set. I could provide more specific mongodb logs, I am currently recording level 4 logs on our DB servers to narrow the problem even further.

**Edit, wrong timestamps in examples**

**Comment by vkarpov15** [comment]
Are these auth failures coming from your node app or from within your replica set?

**Comment by joningib** [comment]
They are from our replica set, but they are also returned from our node app as MongoErrors - "auth failed" as well.
Looks like the issue might be related to:
https://jira.mongodb.org/browse/NODE-719

**Comment by vkarpov15** [comment]
Weird, are you sure every member of your replica set is running the same version?

**Comment by joningib** [comment]
Yes, we are controlling our deployment and mongodb-server core with cloud.mongodb and they are all currently version 3.2.7

**Comment by joningib** [comment]
Alright a small update from our side:
We went into full-force debug mode yesterday and refactored our database connection methodology to our replicaset - along with moving cronjobs from using "cron" into specific resources - we were actually running a separate API app which only ran cronjobs...

I think our main issue is that we started initiating mongoose schema registration on our DB connection object, and initial queries on our main primary data node, before actual db.event('open') was reached - we actually started the specific schema registration procedure on 'fullsetup' which might have been too early.
We now have moved everything to happen first after 'open' from the dbObject and stopped running a separate cron API app.
For the past 24h~ things have been going fine - we haven't registered connection leaks or the authentication error, or SCRAM-SHA-1 badValue.
I'll see how it goes for the next couple of days.

We are now running mongoose 4.5.2, mongodb-core server 3.2.7 and mongodb nodejs driver 2.1.18
